File: backend/src/services/hotspots.ts
import { readFileSync, existsSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { createAdminClient, createClient } from "@insforge/sdk";
import { config, hasInsforge } from "../config.js";
import type { RawHotspot } from "./scoring.js";
import { categoryToSeverity } from "./scoring.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const SEED_PATH = join(__dirname, "../../data/seed-hotspots.json");

let seedCache: RawHotspot[] | null = null;

function getInsforgeClient() {
  if (config.insforge.apiKey) {
    return createAdminClient({ baseUrl: config.insforge.url, apiKey: config.insforge.apiKey });
  }
  return createClient({ baseUrl: config.insforge.url, anonKey: config.insforge.anonKey });
}

function loadSeed(): RawHotspot[] {
  if (seedCache) return seedCache;
  if (!existsSync(SEED_PATH)) {
    seedCache = [];
    return seedCache;
  }
  seedCache = JSON.parse(readFileSync(SEED_PATH, "utf-8")) as RawHotspot[];
  return seedCache;
}

function bboxFilter(lat: number, lng: number, radiusM: number, rows: RawHotspot[]): RawHotspot[] {
  const dLat = radiusM / 111_000;
  const dLng = radiusM / (111_000 * Math.cos((lat * Math.PI) / 180));
  return rows.filter(
    (h) =>
      h.lat >= lat - dLat &&
      h.lat <= lat + dLat &&
      h.lng >= lng - dLng &&
      h.lng <= lng + dLng,
  );
}

export async function queryHotspotsNear(
  lat: number,
  lng: number,
  radiusMeters: number,
  limit = 200,
): Promise<RawHotspot[]> {
  if (hasInsforge()) {
    try {
      const client = getInsforgeClient();
      const dLat = radiusMeters / 111_000;
      const dLng = radiusMeters / (111_000 * Math.cos((lat * Math.PI) / 180));
      const { data, error } = await client.database
        .from("hotspots")
        .select("id,lat,lng,category,severity,occurred_at,source")
        .gte("lat", lat - dLat)
        .lte("lat", lat + dLat)
        .gte("lng", lng - dLng)
        .lte("lng", lng + dLng)
        .limit(limit);

      if (!error && data?.length) {
        return data.map((row: Record<string, unknown>) => ({
          id: String(row.id),
          lat: Number(row.lat),
          lng: Number(row.lng),
          category: String(row.category),
          severity: Number(row.severity) as 1 | 2 | 3 | 4 | 5,
          occurredAt: String(row.occurred_at ?? row.occurredAt),
          source: String(row.source ?? "sfpd"),
        }));
      }
      console.warn("[insforge] hotspots query failed, using seed:", error?.message);
    } catch (e) {
      console.warn("[insforge] hotspots error, using seed:", e);
    }
  }

  const seed = loadSeed();
  return bboxFilter(lat, lng, radiusMeters, seed).slice(0, limit);
}

export async function insertHotspotsBatch(rows: Omit<RawHotspot, "id">[]): Promise<number> {
  if (!hasInsforge()) {
    console.warn("[insforge] not configured — skipping insert");
    return 0;
  }
  const client = getInsforgeClient();
  const payload = rows.map((r) => ({
    lat: r.lat,
    lng: r.lng,
    category: r.category,
    severity: r.severity,
    occurred_at: r.occurredAt,
    source: r.source,
  }));
  const { error } = await client.database.from("hotspots").insert(payload);
  if (error) throw new Error(error.message);
  return payload.length;
}

export { categoryToSeverity, loadSeed };

File: backend/guardian/services/hotspots.py
from __future__ import annotations

import json
from pathlib import Path

from guardian.config import has_insforge
from guardian.services.insforge_client import InsForgeError, admin_client
from guardian.services.scoring import RawHotspot, category_to_severity

_BACKEND_ROOT = Path(__file__).resolve().parents[2]
SEED_PATH = _BACKEND_ROOT / "data" / "seed-hotspots.json"

_seed_cache: list[RawHotspot] | None = None


def load_seed() -> list[RawHotspot]:
    global _seed_cache
    if _seed_cache is not None:
        return _seed_cache
    if not SEED_PATH.exists():
        _seed_cache = []
        return _seed_cache
    _seed_cache = json.loads(SEED_PATH.read_text())
    return _seed_cache


def bbox_filter(lat: float, lng: float, radius_m: float, rows: list[RawHotspot]) -> list[RawHotspot]:
    import math

    d_lat = radius_m / 111_000
    d_lng = radius_m / (111_000 * math.cos(math.radians(lat)))
    return [
        h
        for h in rows
        if lat - d_lat <= h["lat"] <= lat + d_lat and lng - d_lng <= h["lng"] <= lng + d_lng
    ]


def _row_to_hotspot(row: dict) -> RawHotspot:
    return {
        "id": str(row["id"]),
        "lat": float(row["lat"]),
        "lng": float(row["lng"]),
        "category": str(row["category"]),
        "severity": int(row["severity"]),
        "occurredAt": str(row.get("occurred_at") or row.get("occurredAt", "")),
        "source": str(row.get("source", "sfpd")),
    }


async def query_hotspots_near(
    lat: float, lng: float, radius_meters: float, limit: int = 200
) -> list[RawHotspot]:
    if has_insforge():
        try:
            client = admin_client()
            rows = await client.select_hotspots(lat, lng, radius_meters, limit)
            if rows:
                return [_row_to_hotspot(r) for r in rows]
        except InsForgeError as e:
            print(f"[insforge] hotspots query failed, using seed: {e}")
        except Exception as e:
            print(f"[insforge] hotspots error, using seed: {e}")

    seed = load_seed()
    return bbox_filter(lat, lng, radius_meters, seed)[:limit]


async def clear_hotspots_table() -> None:
    if not has_insforge():
        return
    await admin_client().clear_hotspots()


async def insert_hotspots_batch(rows: list[dict]) -> int:
    if not has_insforge():
        return 0
    return await admin_client().insert_hotspots(rows)
